cmd/kubeadm/app/util/users/users_linux.go

```diff
@@ -0,0 +1,602 @@
+// +build linux
+
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package users
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"k8s.io/klog/v2"
+)
+
+// EntryMap holds a map of user or group entries.
+type EntryMap struct {
+	entries map[string]*entry
+}
+
+// entry is a structure that holds information about a UNIX user or group.
+// It partialially conforms parsing of both users from /etc/passwd and groups from /etc/group.
+type entry struct {
+	name      string
+	id        int64
+	gid       int64
+	userNames []string
+	shell     string
+}
+
+// limits is used to hold information about the minimum and maximum system ranges for UID and GID.
+type limits struct {
+	minUID, maxUID, minGID, maxGID int64
+}
+
+const (
+	// these are constants used when parsing /etc/passwd or /etc/group in terms of how many
+	// fields and entry has.
+	totalFieldsGroup = 4
+	totalFieldsUser  = 7
+
+	// klogLevel holds the klog level to use for output.
+	klogLevel = 5
+
+	// noshell holds a path to a binary to disable shell login.
+	noshell = "/bin/false"
+)
+
+var (
+	// these entries hold the users and groups to create as defined in:
+	// https://git.k8s.io/enhancements/keps/sig-cluster-lifecycle/kubeadm/2568-kubeadm-non-root-control-plane
+	usersToCreateSpec = []*entry{
+		{name: "kubeadm-etcd"},
+		{name: "kubeadm-kas"},
+		{name: "kubeadm-kcm"},
+		{name: "kubeadm-ks"},
+	}
+	groupsToCreateSpec = []*entry{
+		{name: "kubeadm-etcd", userNames: []string{"kubeadm-etcd"}},
+		{name: "kubeadm-kas", userNames: []string{"kubeadm-kas"}},
+		{name: "kubeadm-kcm", userNames: []string{"kubeadm-kcm"}},
+		{name: "kubeadm-ks", userNames: []string{"kubeadm-ks"}},
+		{name: "kubeadm-sa-key-readers", userNames: []string{"kubeadm-kas", "kubeadm-kcm"}},
+	}
+
+	// defaultLimits holds the default limits in case values are missing in /etc/login.defs
+	defaultLimits = &limits{minUID: 100, maxUID: 999, minGID: 100, maxGID: 999}
```

Comment:
these are the default ranges for debian and alpine, i did not test other distros.
if they lack /etc/login.defs they are already non-standard - current alpine, older openSUSE (afaik?)

Reply:
Some reference in k/k https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/glusterfs/glusterfs.go#L82-L97

Reply:
these values seem specific to the glusterfs volume plugin.
i have not seen GID limits starting in the 2000 ranges before, but distros are free to pick a range that they like.
as long as the same distro is including these values in /etc/login.defs we should be fine.

Ubuntu for example has the file, but the have commented some of the definitions, which partially defeats the purpose of the file.

```
controlplane $ cat /etc/login.defs |grep GID
GID_MIN                  1000
GID_MAX                 60000
#SYS_GID_MIN              100
#SYS_GID_MAX              999
```

my only hope is that these are defined in pairs and some distro does not have e.g. `SYS_GID_MIN   2000`, but then the `#SYS_GID_MAX ....` line is commented. that would be absurd.